### PR TITLE
ZK-3945: Popup's z-index doesn't work properly

### PR DESCRIPTION
**Pull request by klyvechen**

### CODE REVIEW

File: zktest/src/archive/test2/F85-ZK-3688.zul [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+Copyright (C) 2018 Potix Corporation. All Rights Reserved.
+
+-->
+
+<zk>
+	<hlayout>
+		<div>
+			<label multiline="true">
+				1. Click the button 1 2 3 orderly.
```
Comment (scribetw) [inline zktest/src/archive/test2/F85-ZK-3688.zul]: Why not just a button and execute what you want orderly?

File: zk/src/archive/web/js/zk/widget.js [hunk 2]
```diff
@@ -2161,6 +2162,10 @@ wgt.$f().main.setTitle("foo");
 	 * @return zk.Widget this widget.
 	 */
 	setZIndex: _zkf = function (zIndex, opts) {
+		if (!opts.floatZIndex)
```
Comment (scribetw) [inline zk/src/archive/web/js/zk/widget.js]: Add this opt option to jsdoc

File: zk/src/archive/web/js/zk/widget.js [hunk 3]
```diff
@@ -2161,6 +2162,10 @@ wgt.$f().main.setTitle("foo");
 	 * @return zk.Widget this widget.
 	 */
 	setZIndex: _zkf = function (zIndex, opts) {
+		if (!opts.floatZIndex)
+			this.userZIndex = true;
+		if (!zIndex)
+			delete this.userZIndex;
```
Comment (scribetw) [inline zk/src/archive/web/js/zk/widget.js]: 1. Use `this._userZIndex`
2. `this._userZIndex = false`

File: zk/src/archive/web/js/zk/widget.js [hunk 4]
```diff
@@ -1876,8 +1876,8 @@ wgt.$f().main.setTitle("foo");
 					if (!floating) this.setDomVisible_(node, visible);
 				} else if (visible) {
 					var zi;
-					if (floating)
-						this.setZIndex(zi = _topZIndex(this), {fire: true});
+					if (floating && !this.userZIndex)
```
Comment (scribetw) [inline zk/src/archive/web/js/zk/widget.js]: No need to add `&& !this.userZIndex` here.

In `setZIndex()`, if `opt.floatZIndex && this.userZIndex`, just don't do anything.

File: zk/src/archive/web/js/zk/widget.js (hunk 2 above)
Comment (klyvechen) [reply zk/src/archive/web/js/zk/widget.js]: add,
	/** Sets the Z index.
	 * @param int zIndex the Z index to assign to
	 * @param Map opts if opts.fire is specified, the onZIndex event will be triggered. If opts.floatZIndex is false, 
	 * represent it is not from setFloatZIndex, so the userZIndex mey be true
	 * @return zk.Widget this widget.
	 */

File: zk/src/archive/web/js/zk/widget.js (hunk 3 above)
Comment (klyvechen) [reply zk/src/archive/web/js/zk/widget.js]: ok

File: zk/src/archive/web/js/zk/widget.js (hunk 4 above)
Comment (klyvechen) [reply zk/src/archive/web/js/zk/widget.js]: ok

File: zk/src/archive/web/js/zk/widget.js [hunk 5]
```diff
@@ -2157,10 +2158,16 @@ wgt.$f().main.setTitle("foo");
 	getZindex: _zkf,
 	/** Sets the Z index.
 	 * @param int zIndex the Z index to assign to
-	 * @param Map opts if opts.fire is specified, the onZIndex event will be triggered.
+	 * @param Map opts if opts.fire is specified the onZIndex event will be triggered. If opts.floatZIndex is false, represent it is not from setFloatZIndex, so the userZIndex mey be true.
```
Comment (scribetw) [inline zk/src/archive/web/js/zk/widget.js]: may

**Comment by wenninghsu** [comment]
@zkadmin rebase

**Comment by zkadmin** [comment]
I just pushed up the changes, enjoy!

**Pull request merged**